Fix part of #14033: Added Mypy type annotations to skill_domain.py

dhruvshrivastava opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #14033 .
2. This PR does the following: [Added Mypy type annotations, reduced Mypy type errors to 0]



## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
![Screenshot 2022-01-28 at 4 15 47 AM](https://user-images.githubusercontent.com/28788192/151455867-844ae5d3-0214-4359-9c8d-2aa3ee3fbf04.png)


<!--
Add videos/screenshots of the user-facing interface in various display sizes (mainly phone, tablet, and desktop display size) to demonstrate that the changes made in this PR work correctly. Please also include videos/screenshots of the developer tools browser console, so that we can be sure that there are no console errors.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

dhruvshrivastava committed:
Added Mypy type annotations to skill_domain.py

vojtechjelinek commented on scripts/run_mypy_checks.py:
```diff
@@ -47,6 +47,8 @@
     'core/domain/auth_services_test.py',
     'core/domain/blog_services.py',
     'core/domain/blog_services_test.py',
+    'core/domain/caching_services.py',
+    'core/domain/caching_services_test.py',
```
Don't add anything back, rather fix the errors in _caching_services.py_

dhruvshrivastava replied:
Done.

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -181,11 +185,26 @@ class SkillChange(change_domain.BaseChange):
 
 
+class MisconceptionDict(TypedDict):
+    """Dict for Misconception Domain Object"""
```
```suggestion
    """Dictionary representing the Misconception domain object."""
```

dhruvshrivastava replied:
Done

dhruvshrivastava replied:
Done

dhruvshrivastava replied:
Done

vojtechjelinek replied:
Not fully done. Also ditto in all other TypedDicts in this file.

dhruvshrivastava replied:
Done

vojtechjelinek replied:
Not done.

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -181,11 +185,26 @@ class SkillChange(change_domain.BaseChange):
 
     def __init__(
-            self, misconception_id, name, notes, feedback, must_be_addressed):
+             self,
+             misconception_id: int,
+             name: str,
+             notes: str,
+             feedback: str,
+             must_be_addressed: bool) -> None:
```
```suggestion
         self,
         misconception_id: int,
         name: str,
         notes: str,
         feedback: str,
         must_be_addressed: bool
     ) -> None:
```

dhruvshrivastava replied:
Done

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -221,7 +240,9 @@ def to_dict(self):
 
     @classmethod
-    def from_dict(cls, misconception_dict):
+    def from_dict(
+        cls, misconception_dict: MisconceptionDict
+        ) -> Misconception:
```
```suggestion
    def from_dict(
        cls, misconception_dict: MisconceptionDict
    ) -> Misconception:
```

dhruvshrivastava replied:
Done

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -289,10 +310,17 @@ def validate(self):
 
 
+class RubricDict(TypedDict):
+    """Dict for Rubric Domain Object"""
```
```suggestion
    """Dictionary representing the Rubric domain object."""
```

dhruvshrivastava replied:
Done

dhruvshrivastava replied:
Done

vojtechjelinek replied:
Not fully done.

dhruvshrivastava replied:
Done

vojtechjelinek replied:
Not done.

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -360,7 +388,11 @@ def validate(self):
     """Domain object for representing the worked_example dict."""
 
-    def __init__(self, question, explanation):
+    def __init__(
+                self,
+                question: state_domain.SubtitledHtml,
+                explanation: state_domain.SubtitledHtml
+                ) -> None:
```
```suggestion
    def __init__(
        self,
        question: state_domain.SubtitledHtml,
        explanation: state_domain.SubtitledHtml
    ) -> None:
```

vojtechjelinek replied:
ditto below

dhruvshrivastava replied:
Done

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -412,25 +448,37 @@ def from_dict(cls, worked_example_dict):
 
 
+class SkillContentsDict(TypedDict):
+    """Dict for Skill Contents Domain Object"""
```
rephrase as above

dhruvshrivastava replied:
Done

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -513,31 +563,83 @@ def from_dict(cls, skill_contents_dict):
 
 
+class SkillDict(TypedDict):
+    """Dict for Skill Domain Object"""
```
Rephrase as above.

vojtechjelinek replied:
ditto below

dhruvshrivastava replied:
Done

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -513,31 +563,83 @@ def from_dict(cls, skill_contents_dict):
+    language_code: str
+    version: int
+    next_misconception_id: int
+    superseding_skill_id: str|None
```
```suggestion
    superseding_skill_id: Optional[str]
```

dhruvshrivastava replied:
Done

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -513,31 +563,83 @@ def from_dict(cls, skill_contents_dict):
+            language_code: str,
+            version: int,
+            next_misconception_id: int,
+            superseding_skill_id: str|None,
```
```suggestion
            superseding_skill_id: Optional[str],
```

dhruvshrivastava replied:
Done

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -591,7 +693,9 @@ def __init__(
 
     @classmethod
-    def require_valid_skill_id(cls, skill_id):
+    def require_valid_skill_id(
+        cls, skill_id: str
+        ) -> None:
```
```suggestion
    def require_valid_skill_id(
        cls, skill_id: str
    ) -> None:
```

ditto below

dhruvshrivastava replied:
Done.

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -802,19 +908,21 @@ def serialize(self):
         # to_dict().
         skill_dict['version'] = self.version
-
+        # Using type ignore[misc] below because to_disc()
+        # creates a dict with no created_on and last_updated keys.
```
I don't understand this.

dhruvshrivastava replied:
Hi, sorry for the unclear comment. The to_dict() function of the Skill class returns a dictionary with no 'created_on' and 'last_updated' keys so that throws the 'missing keys' as shown below:
![Screenshot 2022-01-28 at 10 34 38 PM](https://user-images.githubusercontent.com/28788192/151590226-5e1b6d58-82fb-48f2-877f-5c7abb8ce313.png)
 I tried various workarounds but couldn't resolve the problem so I decided to use a type ignore. Is there something else I could do to resolve the issue?

dhruvshrivastava replied:
Sorry for the unclear comment. I have made the changes and resolved the issue instead of using a type ignore.

dhruvshrivastava replied:
Done

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -802,19 +908,21 @@ def serialize(self):
 
     @classmethod
-    def deserialize(cls, json_string):
+    def deserialize(
+        cls, json_string: str) -> Skill:
```
```suggestion
    def deserialize(cls, json_string: str) -> Skill:
```

dhruvshrivastava replied:
Done

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -844,8 +952,12 @@ def deserialize(cls, json_string):
+            cls,
+            skill_dict: SkillDict,
+            skill_version: int=0,
+            skill_created_on: datetime.datetime|None=None,
+            skill_last_updated: datetime.datetime|None=None
```
use optional and fix formatting

dhruvshrivastava replied:
Done

vojtechjelinek reviewed (changes requested):
Thanks! Took a partial pass, please fix the formatting of the code so that it look similar to other typed domain files.

dhruvshrivastava committed:
Fix formatting and Mypy errors in skill_domain.py, caching_services.py and caching_services_test.py

dhruvshrivastava commented:
Hi @vojtechjelinek, I've made the changes. Can you PTAL? thanks

dhruvshrivastava committed:
Update skill_domain.py

vojtechjelinek commented:
@dhruvshrivastava Please make sure to reply to all addressed comments with "Done".

dhruvshrivastava commented:
> @dhruvshrivastava Please make sure to reply to all addressed comments with "Done".

Done! thanks

vojtechjelinek commented on core/domain/caching_services_test.py:
```diff
@@ -840,7 +840,7 @@ def test_skills_with_unicode_characters_are_set_and_get_correctly(
         self.assertEqual(
             default_skill.to_dict(),
-            skills[skill_id].to_dict()) # type: ignore[no-untyped-call]
+            skills[skill_id].to_dict())
```
Can be moved to previous line.

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -17,15 +17,20 @@
+import datetime
 import json
 
+
```
unneeded empty line

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -382,26 +416,30 @@ def validate(self):
+        self.explanation.validate() # type: ignore[no-untyped-call]
 
-    def to_dict(self):
+    def to_dict(self) -> Dict[str, Dict[str, state_domain.SubtitledHtml]]:
```
Use `TypedDict`

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -382,26 +416,30 @@ def validate(self):
-    def from_dict(cls, worked_example_dict):
+    def from_dict(
+        cls,
+        worked_example_dict: Dict[str,
+        Dict[str, state_domain.SubtitledHtml]]
```
Use `TypedDict`

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -844,8 +951,12 @@ def deserialize(cls, json_string):
-            skill_last_updated=None):
+        cls,
+        skill_dict: SkillDict,
+        skill_version: int=0,
+        skill_created_on: Optional[datetime.datetime]=None,
+        skill_last_updated: Optional[datetime.datetime]=None
```
```suggestion
        skill_version: int = 0,
        skill_created_on: Optional[datetime.datetime] = None,
        skill_last_updated: Optional[datetime.datetime] =None 
```

ditto elsewhere

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -938,7 +1054,11 @@ def generate_skill_misconception_id(self, misconception_id):
-            cls, skill_contents_dict, conversion_fn):
+        cls,
+        skill_contents_dict: SkillContentsDict,
+        conversion_fn: Callable[[state_domain.SubtitledHtml],
+        state_domain.SubtitledHtml]
```
```suggestion
        conversion_fn: Callable[
            [state_domain.SubtitledHtml], state_domain.SubtitledHtml
        ]
```

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -1148,13 +1298,20 @@ def _convert_rubric_v1_dict_to_v2_dict(cls, rubric_dict):
         """
-        explanation = rubric_dict['explanation']
-        del rubric_dict['explanation']
+        # Using type ignore[misc] below because the Rubric domain object
+        # contains an explanations field instead of explanation
+        # Hence RubricDict has an explanations key instead of explanation
+        # but the below code is using explanation.
+        explanation = rubric_dict['explanation'] # type: ignore[misc]
+        del rubric_dict['explanation'] # type: ignore[misc]
```
Ah, interesting, so the previous schema of rubric contained `explanation` but now it contains `explanations`, not sure if this is the right solution, but I will have to think about it.

dhruvshrivastava replied:
Yep. Is there any alternative fix for this or should I leave this as it is for now?

dhruvshrivastava replied:
@vojtechjelinek Could you please guide me on how to resolve this issue?

vojtechjelinek replied:
Lets just keep the ignores.

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -1130,15 +1277,18 @@ def _convert_misconception_v4_dict_to_v5_dict(cls, misconception_dict):
 
     @classmethod
-    def _convert_rubric_v1_dict_to_v2_dict(cls, rubric_dict):
+    def _convert_rubric_v1_dict_to_v2_dict(
+        cls,
+        rubric_dict: RubricDict
+    ) -> RubricDict:
```
```suggestion
    def _convert_rubric_v1_dict_to_v2_dict(
        cls, rubric_dict: RubricDict
    ) -> RubricDict:
```

ditto elsewhere

dhruvshrivastava replied:
Done

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -1628,16 +1850,40 @@ def to_dict(self):
 
 
+class AugmentedSkillSummaryDict(TypedDict):
+    """Dict for Augmented Skill Summary Domain Object"""
```
Not done.

dhruvshrivastava replied:
Done

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

dhruvshrivastava committed:
Update skill_domain.py and caching_services_test.py

aks681 commented on core/domain/skill_domain.py:
```diff
@@ -412,25 +454,38 @@ def from_dict(cls, worked_example_dict):
+class SkillContentsDict(TypedDict):
+    """Dictionary representing the Skill Contents Domain Object"""
+
+    explanation: Dict[str, state_domain.SubtitledHtml]
+    worked_examples: List[WorkedExampleDict]
+    recorded_voiceovers: Dict[str, Dict[str, state_domain.RecordedVoiceovers]]
+    written_translations: Dict[str, Dict[str, state_domain.WrittenTranslations]]
```
For the dicts, aren't the values also a dictionary and not the domain objects? (So, the dict form of SubtitledHtml instead of a SubtitledHtml object?)

dhruvshrivastava replied:
@aks681 That'd mean creating separate `TypedDicts` for the `state_domain` domain objects, which'd be done once type annotations are added to `state_domain` or is there any alternative fix for this?

vojtechjelinek replied:
Hmmm, @aks681 is right don't use `RecordedVoiceovers`, you will either need to wait for `state_domain` to be typed or add the `TypedDicts` yourself.

aks681 reviewed (commented):
Just one comment, else lgtm as codeowner

vojtechjelinek commented:
@dhruvshrivastava This comment is not addressed everywhere https://github.com/oppia/oppia/pull/14794#discussion_r794605998

dhruvshrivastava commented:
> @dhruvshrivastava This comment is not addressed everywhere [#14794 (comment)](https://github.com/oppia/oppia/pull/14794#discussion_r794605998)

I have addressed them now. Thanks

dhruvshrivastava committed:
Merge branch 'oppia:develop' into dhruv-oppia

dhruvshrivastava committed:
Merge branch 'oppia:develop' into dhruv-oppia

dhruvshrivastava committed:
Merge branch 'oppia:develop' into dhruv-oppia

Commit:
Add type annotations to skill_domain_test

vojtechjelinek commented:
@dhruvshrivastava Some of the old comments are still not fully addressed.

dhruvshrivastava commented:
@vojtechjelinek I'm a little confused, I have addressed the comments and committed the changes from my end. For example, I resolved the docstring issue (https://github.com/oppia/oppia/pull/14794#discussion_r794605998) and it shows up in my commits as well. Screenshot attached below. 
![Screenshot 2022-02-18 at 2 08 46 AM](https://user-images.githubusercontent.com/28788192/154567119-247132ce-5865-4a51-b0ac-3289f8ac9e5d.png)

Am I missing something?

vojtechjelinek commented:
@dhruvshrivastava But my suggestion is also changing the uppercase D and O to lowercase 
![image](https://user-images.githubusercontent.com/10142938/154664869-56d4007a-ca29-43a4-a80b-38e56c1a516a.png)

Commit:
Fix formatting in skill_domain

dhruvshrivastava commented:
> @dhruvshrivastava But my suggestion is also changing the uppercase D and O to lowercase ![image](https://user-images.githubusercontent.com/10142938/154664869-56d4007a-ca29-43a4-a80b-38e56c1a516a.png)

Oh my bad, I completely missed that. Fixed it now. @vojtechjelinek PTAL

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -181,11 +185,27 @@ class SkillChange(change_domain.BaseChange):
 
 
+class MisconceptionDict(TypedDict):
+    """Dictionary representing the Misconception domain object"""
```
```suggestion
    """Dictionary representing the Misconception domain object."""
```

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -289,10 +311,17 @@ def validate(self):
 
 
+class RubricDict(TypedDict):
+    """Dictionary representing the Rubric domain object"""
```
```suggestion
    """Dictionary representing the Rubric domain object."""
```

ditto elsewhere

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -357,10 +386,21 @@ def validate(self):
                     explanation)
 
 
+class WorkedExampleDict(TypedDict):
+    """Dictionary representing the WorkedExample domain object"""
+
+    question: Dict[str, state_domain.SubtitledHtml]
+    explanation: Dict[str, state_domain.SubtitledHtml]
```
Should this be placed into _state_domain.py_

vojtechjelinek replied:
ditto for some other dicts, if the domain object is in _state_domain.py_ then the dict should be there too.

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -513,31 +570,85 @@ def from_dict(cls, skill_contents_dict):
         return skill_contents
 
 
+class SkillDict(TypedDict, total=False):
```
Why `total=False`?

vojtechjelinek commented on core/domain/skill_domain.py:
```diff
@@ -1258,31 +1418,34 @@ def get_all_html_content_strings(self):
 
-    def record_that_all_questions_are_merged(self, all_questions_merged):
+    def record_that_all_questions_are_merged(
+        self,
+        all_questions_merged: bool
```
```suggestion
        self, all_questions_merged: bool
```

vojtechjelinek replied:
ditto below

vojtechjelinek commented on core/domain/skill_domain_test.py:
```diff
@@ -69,33 +69,42 @@ def setUp(self):
         )
+    # We have ignored [override] here because the signature of this method
+    # doesn't match with TestBase.
 
```
Remove this empty line.

vojtechjelinek commented on core/domain/skill_domain_test.py:
```diff
@@ -69,33 +69,42 @@ def setUp(self):
+        with self.assertRaisesRegex(    # type: ignore[no-untyped-call]
             utils.ValidationError, 'Skill id should be a string'):
-            skill_domain.Skill.require_valid_skill_id(10)
+            skill_domain.Skill.require_valid_skill_id(10) # type: ignore[arg-type]
```
I think you can remove these tests for the wrong type.

vojtechjelinek commented on core/domain/skill_domain_test.py:
```diff
@@ -713,7 +782,7 @@ def test_skill_change_object_with_migrate_misconceptions_schema_to_latest_versio
 
     def test_skill_change_object_with_migrate_rubrics_schema_to_latest_version(
-            self):
+            self) -> None:
```
```suggestion
        self
    ) -> None:
```

ditto elsewhere

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.
